File: _generic/primes_helper.py
# Generic Primes

if __name__ == '__main__':

    try:

        # Generate an 0-indexed array where:
        # sieve[i] = False indicates non-prime
        # sieve[i] = True indicates prime
        def sieve_of_eratosthenes(num):
            sieve = [0] * num
            sieve[0] = False
            # print(sieve)

            for x in range(2, num+1, 1):
                if sieve[x-1] is 0:
                    sieve[x-1] = True

                for y in range(x*2, num+1, x):
                    sieve[y-1] = False

                # print(sieve)

            print(sieve)
            return sieve

        def prime_factorization(num):
            # Offset num by one when accessing
            primes = sieve_of_eratosthenes(num)
            result = []
            rem = num

            while rem % 2 == 0:
                rem = rem / 2
                result.append(2)

            # Must be odd or not divisible by 2
            for x in range(3, num+1, 2):
                if (primes[x-1] == False):
                    continue

                while (rem % x == 0):
                    result.append(x)
                    rem = rem / x

                if rem == 1 or rem == 0:
                    break

            print(result)
            return result

        def number_of_divisors(num):
            prime_factors = prime_factorization(num)
            result = 1
            hm = {}

            for x in range(0, len(prime_factors), 1):
                n = prime_factors[x]

                if not hm.get(n) is None:
                    hm[n] = hm[n] + 1
                else:
                    hm[n] = 1

            keys = hm.keys()
            
            for x in range(0, len(keys), 1):
                n = hm[keys[x]]
                result = result * (n + 1)

            print(result)
            return result

        #sieve_of_eratosthenes(10)
        #sieve_of_eratosthenes(20)
        #sieve_of_eratosthenes(50)

        #prime_factorization(12)
        #prime_factorization(18)
        #prime_factorization(500)

        #number_of_divisors(24)
        #number_of_divisors(18)
        #number_of_divisors(10)
        #number_of_divisors(5)

    except Exception as ex:

        print('Exception: ' + str(ex))

File: _generic/primes_helper.js
/*
  Generate an 0-indexed array where:
  sieve[i] = False indicates non-prime
  sieve[i] = True indicates prime
*/

const sieveOfEratosthenes = (num) => {
  let sieve = [];
  for (let i = 0; i < num; i++) {
    sieve.push(0);
  }

  sieve[0] = false;
  console.log(sieve);

  for (let i = 2; i < num + 1; i++) {
    if (sieve[i - 1] === 0) sieve[i - 1] = true;

    for (let j = i * 2; j < num + 1; j += i) {
      sieve[j - 1] = false;
    }
  }

  console.log(sieve);
  return sieve;
};

const primeMap = (arr) => {
  let result = {};

  for (let i = 0; i < arr.length; i++) {
    if (arr[i]) result[arr[i] + 1] = true;
  }

  console.log(result);
  return result;
};

const primeFactorization = (num) => {
  let primes = sieveOfEratosthenes(num);
  let result = [];
  let rem = num;

  while (rem % 2 === 0) {
    rem = rem / 2;
    result.push(2);
  }

  for (let i = 3; i < num + 1; i += 2) {
    if (primes[i - 1] === false) continue;

    while (rem % i === 0) {
      result.push(i);
      rem = rem / i;
    }

    if (rem === 1 || rem === 0) break;
  }

  console.log(result);
  return result;
};

const numberOfDivisors = (num) => {
  let primeFactors = primeFactorization(num);
  let result = 1;
  let hm = {};

  for (let i = 0; i < primeFactors.length; i++) {
    const N = primeFactors[i];

    if (hm[N]) hm[N]++;
    else hm[N] = 1;
  }

  const KEYS = Object.keys(hm);

  for (let i = 0; i < KEYS.length; i++) {
    const N = hm[KEYS[i]];
    result *= N + 1;
  }

  console.log(result);
  return result;
};

const checkPrime = (num) => {
  const SQ_RT = Math.ceil(Math.pow(num, 1 / 2));

  let isEven = false;
  if (num % 2 === 0) isEven = true;

  for (let i = 2; i <= SQ_RT; i++) {
    if (num % i === 0) return false;
    if (isEven) i++;
  }

  return true;
};
